src/core/lib/iomgr/work_serializer.cc

```diff
@@ -55,17 +56,20 @@ void WorkSerializer::WorkSerializerImpl::Run(
     gpr_log(GPR_INFO, "WorkSerializer::Run() %p Scheduling callback [%s:%d]",
             this, location.file(), location.line());
   }
-  const size_t prev_size = size_.fetch_add(1);
+  const size_t prev_size = size_.fetch_add(1, std::memory_order_acq_rel);
   // The work serializer should not have been orphaned.
   GPR_DEBUG_ASSERT(prev_size > 0);
-  if (prev_size == 1) {
-    // There is no other closure executing right now on this work serializer.
-    // Execute this closure immediately.
+  // If there is no other callback executing right now on this work serializer,
+  // try to grab the lock and execute the callback immediately.
+  bool expected = false;
+  if (prev_size == 1 && draining_.compare_exchange_strong(expected, true)) {
```

Comment:
I'm really wondering now whether you need `seq_cst`.  In several locations, there are checks on both `draining_` and `size_`. 

Do you think you can put them both in the same variable? For example, using the most significant bit of size_ for draining?

Reply:
That is similar to Mark's comment. Done!